arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java

```diff
@@ -252,17 +297,10 @@ public boolean hasNext() {
         }
       } catch (IOException | RuntimeException e) {
         if (currentTask != null && !currentTask.isDataTask()) {
-          throw new RuntimeException(
```

Comment:
while reviewing this part of the PR, I noticed that `hasNext` isn't idempotent, meaning that if you call `hasNext` multiple times, it will actually advance the pointer and IMO this kind of logic should only happen inside `next` and not `hasNext`.

I also wonder why we need to catch exceptions in a `hasNext` method, so maybe this is something that can be improved in a follow-up?

Reply:
Good catch. I'll make hasNext() idempotent in a follow-up cr. 

On the point of exceptions, I think exceptions should not cause iteration to silently stop and it is better to raise them. There are other instances in Iceberg where similar exception handling is done, for example: https://github.com/apache/iceberg/blob/90225d6c9413016d611e2ce5eff37db1bc1b4fc5/api/src/main/java/org/apache/iceberg/io/CloseableIterable.java#L156. Let me know if I understand your comment correctly. I'm happy to discuss this further and improve the code accordingly.

Reply:
I was just curious whether this kind of exception handling was necessary here or not. I guess once `hasNext` is idempotent, the exception handling (if necessary) can be moved to `next`

Reply:
The change will be the following:
`--- a/arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java
+++ b/arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java
@@ -25,6 +25,7 @@ import java.util.Collection;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.NoSuchElementException;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
@@ -172,7 +173,6 @@ public class ArrowReader extends CloseableGroup {
     private final int batchSize;
     private final boolean reuseContainers;
     private CloseableIterator<ColumnarBatch> currentIterator;
-    private ColumnarBatch current;
     private FileScanTask currentTask;
 
     /**
@@ -239,7 +239,6 @@ public class ArrowReader extends CloseableGroup {
       try {
         while (true) {
           if (currentIterator.hasNext()) {
-            this.current = currentIterator.next();
             return true;
           } else if (fileItr.hasNext()) {
             this.currentIterator.close();
@@ -268,7 +267,11 @@ public class ArrowReader extends CloseableGroup {
 
     @Override
     public ColumnarBatch next() {
-      return current;
+      if (currentIterator.hasNext()) {
+        return currentIterator.next();
+      } else {
+        throw new NoSuchElementException();
+      }
     }`
The exception handling will not change. This is similar to he CloseableIterable.hasNext().